File: ruby/bird-count/bird_count.rb
class BirdCount
  BUSY_DAYS = 5

  attr_reader :history

  def self.last_week
    [0, 2, 5, 3, 7, 8, 4]
  end

  def initialize(birds_per_day)
    @history = birds_per_day
  end

  def yesterday
    history[-2]
  end

  def total
    history.sum
  end

  def busy_days
    history.select { |birds| birds >= BUSY_DAYS }.count
  end

  def day_without_birds?
    history.any?(&:zero?)
  end
end

File: elixir/bird-count/lib/bird_count.ex
defmodule BirdCount do
  def today([]), do: nil

  def today([count | _]), do: count

  def increment_day_count([]), do: [1]
  def increment_day_count([count | rest]), do: [count + 1 | rest]

  def has_day_without_birds?([]), do: false
  def has_day_without_birds?([0 | _]), do: true
  def has_day_without_birds?([_ | rest]), do: has_day_without_birds?(rest)

  def total([]), do: 0
  def total([acc]), do: acc
  def total([acc, next | rest]), do: total([acc + next | rest])

  def busy_days([]), do: 0

  def busy_days(list), do: busy_days(list, 0)

  def busy_days([], acc), do: acc
  def busy_days([day | rest], acc) when day >= 5, do: busy_days(rest, acc + 1)
  def busy_days([_ | rest], acc), do: busy_days(rest, acc)
end
